server/src/main/java/org/apache/druid/server/coordination/SegmentLoadDropHandler.java

```diff
@@ -216,21 +208,30 @@ void removeSegment(
     SegmentChangeStatus result = null;
     try {
       announcer.unannounceSegment(segment);
-      segmentsToDelete.add(segment);
+
+      final SegmentDropLatch dropLatch = new SegmentDropLatch();
+      final SegmentDropLatch priorLatch = segmentDropLatches.putIfAbsent(segment, dropLatch);
+
+      if (priorLatch != null) {
+        log.warn("Cannot drop segment[%s] that already has a drop pending. Ignoring.", segment.getId());
+        return;
+      }
 
       Runnable runnable = () -> {
         try {
-          synchronized (segmentDeleteLock) {
-            if (segmentsToDelete.remove(segment)) {
-              segmentManager.dropSegment(segment);
-            }
+          if (dropLatch.startDropping()) {
+            segmentManager.dropSegment(segment);
           }
         }
         catch (Exception e) {
           log.makeAlert(e, "Failed to remove segment! Possible resource leak!")
              .addData("segment", segment)
              .emit();
         }
+        finally {
+          dropLatch.doneDropping();
+          segmentDropLatches.remove(segment, dropLatch);
```

Comment:
Based on the changes made to `addSegment` and `removeSegment`, I think we can get away with calling `Map.remove(key)` here instead of `Map.remove(key, value)`.